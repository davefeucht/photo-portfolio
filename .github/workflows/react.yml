# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: React CI/CD

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x]

    steps:
    - name: Checkout code
      uses: actions/checkout@v2

    - name: Install Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}

    - name: Enable corepack
      run: corepack enable

    - name: Install dependencies
      run: yarn install

    - name: Build project
      run: npm run build --if-present

    - name: Run tests
      run: npm test
