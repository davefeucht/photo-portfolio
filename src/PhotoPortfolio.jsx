/*******************
* Top level PhotoPortfolio component. Sets up the base site URL and 
* loads the title bar and list of categories.
*******************/

import React from "react";
import TitleBar from "./TitleBar.jsx";
import Categories from "./Categories.jsx";
import Footer from "./Footer.jsx";

export default class PhotoPortfolio extends React.Component {
  constructor(props) {
    super(props);

    //Set up the base site url
    this.site_url = "throughapinhole.com";

  }

  render () {

    return (
      <div className="app">
        <TitleBar />
        <div className="content">
          <Categories site={this.site_url} />
        </div>
        <Footer />
      </div>
    );
  }
}