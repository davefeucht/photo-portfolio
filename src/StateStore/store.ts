import { configure, makeAutoObservable } from 'mobx';
import { ImageData, Image, Store, Category, Post, Page } from '../utils/types';

configure({ enforceActions: 'observed' });

class stateStore implements Store {
    menuState = 'closed'
    applicationRoot: HTMLElement = null;
    screenInfo: {
        width: number,
        height: number
    }
    siteInfo: {
        siteName: string,
        siteUrl: string
    }
    visiblePost: {
        postId: number,
        postTitle: string,
        tags: number[],
        tagNames: string[],
        fullImageUrl: string,
        width: number,
        height: number
    };
    categoryList: Category[] = [];
    currentCategoryPosts: Post[] = [];
    currentCategoryData: Category = {
        id: 0,
        count: 0,
        description: '',
        link: '',
        name: '',
        slug: '',
        taxonomy: '',
        parent: 0
    };
    pages: Page[] = [];
    currentPageData: Page = {
        id: 0,
        link: '',
        modified: '',
        slug: '',
        type: '',
        title: {
            rendered: ''
        },
        content: {
            rendered: '',
            protected: false
        },
        author: 0,
        featured_media: 0
    };
    maxItemsPerPage = 10;

    constructor() {
        makeAutoObservable(this);
    }

    setMenuState = (state: string) => {
        this.menuState = state;
    }

    toggleMenuState = () => {
        this.menuState = this.menuState === 'closed' ? 'open' : 'closed';
    }

    setApplicationRoot = (element: HTMLElement) => {
        this.applicationRoot = element;
    }

    setSiteName = (name: string) => {
        this.siteInfo.siteName = name;
    }

    setVisiblePost = (postId: number, postTitle: string) => {
        this.visiblePost.postId = postId;
        this.visiblePost.postTitle = postTitle;
    }

    setVisiblePostImage = (fullImageUrl: string) => {
        this.visiblePost.fullImageUrl = fullImageUrl;
    }

    setVisiblePostTags = (tagNames: string[]) => {
        tagNames.forEach(tagName => {
            this.visiblePost.tagNames.push(tagName);
        });
    }

    clearVisiblePostTagNames = () => {
        this.visiblePost.tagNames.splice(0, this.visiblePost.tagNames.length);
    }

    setCategoryList = (categories: Category[]) => {
        this.categoryList.length = 0;
        categories.forEach((category, index) => {
            this.categoryList[index] = category;
        })
    }

    setCategoryPosts = (posts: Post[]) => {
        this.currentCategoryPosts.length = 0;
        posts.forEach((post, index) => {
            this.currentCategoryPosts[index] = post;
        })
    }

    setCategoryData = (categoryData: Category) => {
        Object.keys(categoryData).forEach((key: keyof Category) => {
            this.currentCategoryData[key] = categoryData[key];
        });
    }

    setThumbnailImageUrl = (imageData: ImageData) => {
        this.currentCategoryPosts[imageData.post_index].thumbnail_image = imageData.image_url;
    }

    setCurrentPost = (postData: Post) => {
        Object.keys(postData).forEach((key: keyof Image) => {
            this.visiblePost[key] = postData[key];
        })
    }

    setPages = (pages: Page[]) => {
        pages.forEach(page => {
            this.pages.push(page);
        });
    }

    setPageData = (pageData: Page) => {
        Object.assign(this.currentPageData, pageData);
    }
}

export default stateStore;
