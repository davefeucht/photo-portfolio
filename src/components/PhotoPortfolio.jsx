/*******************
* Top level PhotoPortfolio component. Sets up the base site URL and 
* loads the title bar and list of categories.
*******************/

import React from "react";
import axios from "axios";
import App from "./styledComponents/App.jsx";
import TitleBar from "./TitleBar.jsx";
import MainContent from "./styledComponents/MainContent.jsx";
import Categories from "./Categories.jsx";
import Category from "./Category.jsx";
import Footer from "./Footer.jsx";

export default class PhotoPortfolio extends React.Component {

  static site_url = "throughapinhole.com";

  state = {
    siteName: "",
    showAllCategories: true,
    singleCategoryToShow: {id: 0, name: ""},
  }

  _getSiteInformation() {
    const getSiteInformationURI = `https://${PhotoPortfolio.site_url}/wp-json/`;

    axios.get(getSiteInformationURI)
      .then(response => {
        const siteName = response.data.name;
        this.setState({siteName});
      })
      .catch(error => {
        console.log(error.message);
      })
  }

  //Method to display a single category when one is clicked on.
  _showSpecificCategory(categoryId, categoryName) {
    this.setState({showAllCategories: false}); 
    this.setState({singleCategoryToShow: {id: categoryId, name: categoryName}}); 
  }
  
  //Method to display all existing categories.
  _showAllCategories() {
    this.setState({showAllCategories: true});
  }

  componentWillMount() {
    this._getSiteInformation();
  }

  render () {
    let contentToDisplay = <Categories site={PhotoPortfolio.site_url} showSingleCategory={this._showSpecificCategory.bind(this)} />

    if(!this.state.showAllCategories) {
      contentToDisplay = <Category key={this.state.singleCategoryToShow.id.toString()} id={this.state.singleCategoryToShow.id} name={this.state.singleCategoryToShow.name} site={PhotoPortfolio.site_url} clickCategory={this._showAllCategories.bind(this)} />; 
    }

    return (

      <App>
        <TitleBar siteName={this.state.siteName} />
        <MainContent>
          {contentToDisplay}
        </MainContent>
        <Footer />
      </App>
    );
  }
}