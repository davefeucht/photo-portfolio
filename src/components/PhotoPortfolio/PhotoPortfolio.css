.photo-portfolio {
  grid-area: content;
}

.app {
  height: 100vh;
  display: grid;
  grid-template-areas: "header" "content" "footer";
  grid-template-rows: 3rem auto 2rem;
}