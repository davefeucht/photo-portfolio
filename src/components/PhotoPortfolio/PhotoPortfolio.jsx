/** *****************
* Top level PhotoPortfolio component. Sets up the base site URL and 
* loads the title bar and list of categories.
****************** */

import './PhotoPortfolio.css';

import { reaction, runInAction } from 'mobx';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import {
    Route,
    Routes,
    useParams
} from 'react-router-dom';

import { getCategories, getPages } from '../../utils/Api';
import Categories from '../Categories/Categories.jsx';
import Category from '../Category/Category.jsx';
import Footer from '../Footer/Footer.jsx';
import Menu from '../Menu/Menu.jsx';
import Page from '../Page/Page.jsx';
import Post from '../Post/Post.jsx';
import TitleBar from '../TitleBar/TitleBar.jsx';

function PhotoPortfolio({ stateStore }) {
    const { categoryId, pageId } = useParams();

    const setScreenSize = () => {
        runInAction(() => {
            stateStore.screenInfo.width = stateStore.applicationRoot.clientWidth;
            stateStore.screenInfo.height = stateStore.applicationRoot.clientHeight;
        });
    };

    const setColumns = (width, height) => {
        if ((width < height)) {
            document.body.style.setProperty('--number-of-columns', 2);
        } else {
            document.body.style.setProperty('--number-of-columns', 5);
        }
    };

    useEffect(() => {
        window.addEventListener('resize', setScreenSize.bind(this));
        getCategories(stateStore.siteInfo.siteUrl)
            .then(categories => {
                stateStore.setCategoryList(categories);
            });
        getPages(stateStore.siteInfo.siteUrl)
            .then(pages => {
                stateStore.setPages(pages);
            });

        setScreenSize();
        setColumns(stateStore.screenInfo.width, stateStore.screenInfo.height);
        const disposer = reaction(
            () => [stateStore.applicationRoot.clientWidth, stateStore.applicationRoot.clientHeight],
            () => setScreenSize(stateStore.applicationRoot.clientWidth, stateStore.applicationRoot.clientHeight)
        );
        const disposer2 = reaction(
            () => [stateStore.screenInfo.width, stateStore.screenInfo.height],
            () => setColumns(stateStore.screenInfo.width, stateStore.screenInfo.height)
        );
        return () => {
            window.removeEventListener('resize', setScreenSize);
            disposer();
            disposer2();
        };
    }, []);

    return (
        <div className="app">
            <TitleBar stateStore={stateStore} />
            <div className="photo-portfolio">
                <Routes>
                    <Route
                        path="/"
                        element={<Categories stateStore={stateStore} />}
                    />
                    <Route
                        path="page/:pageId"
                        element={<Page key={pageId} stateStore={stateStore} />}
                    />
                    <Route
                        path="category/:categoryId"
                        element={<Category key={categoryId} stateStore={stateStore} />}
                    >
                        <Route
                            path="post/:postId"
                            element={<Post stateStore={stateStore} />}
                        />
                    </Route>
                </Routes>
            </div>
            <Menu stateStore={stateStore} />
            <Footer />
        </div>
    );
}

PhotoPortfolio.displayName = 'PhotoPortfolio';

PhotoPortfolio.propTypes = {
    stateStore: PropTypes.object.isRequired
};

export default observer(PhotoPortfolio);
