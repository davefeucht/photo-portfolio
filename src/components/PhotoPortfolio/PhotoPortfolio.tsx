/** *****************
* Top level PhotoPortfolio component. Sets up the base site URL and 
* loads the title bar and list of categories.
****************** */

import './PhotoPortfolio.css';

import { reaction, runInAction } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect } from 'react';
import {
    Route,
    Routes,
    useParams
} from 'react-router-dom';

import { getCategories, getPages } from '../../utils/Api';
import { Store } from '../../utils/types';
import Categories from '../Categories/Categories';
import Category from '../Category/Category';
import Footer from '../Footer/Footer';
import Menu from '../Menu/Menu';
import Page from '../Page/Page';
import Post from '../Post/Post';
import TitleBar from '../TitleBar/TitleBar';

interface PhotoPortfolioProps {
    stateStore: Store
}

const PhotoPortfolio: React.FC<PhotoPortfolioProps> = ({ stateStore }) => {
    const { categoryId, pageId } = useParams();

    const setScreenSize = () => {
        runInAction(() => {
            stateStore.screenInfo.width = stateStore.applicationRoot.clientWidth;
            stateStore.screenInfo.height = stateStore.applicationRoot.clientHeight;
        });
    };

    const setColumns = (width: number, height: number) => {
        if ((width < height)) {
            document.body.style.setProperty('--number-of-columns', '2');
        } else {
            document.body.style.setProperty('--number-of-columns', '5');
        }
    };

    useEffect(() => {
        window.addEventListener('resize', setScreenSize.bind(this));
        getCategories(stateStore.siteInfo.siteUrl)
            .then(categories => {
                stateStore.setCategoryList(categories);
            });
        getPages(stateStore.siteInfo.siteUrl)
            .then(pages => {
                stateStore.setPages(pages);
            });

        setScreenSize();
        setColumns(stateStore.screenInfo.width, stateStore.screenInfo.height);
        const disposer = reaction(
            () => [stateStore.applicationRoot.clientWidth, stateStore.applicationRoot.clientHeight],
            () => setScreenSize()
        );
        const disposer2 = reaction(
            () => [stateStore.screenInfo.width, stateStore.screenInfo.height],
            () => setColumns(stateStore.screenInfo.width, stateStore.screenInfo.height)
        );
        return () => {
            window.removeEventListener('resize', setScreenSize);
            disposer();
            disposer2();
        };
    }, []);

    return (
        <div className="app">
            <TitleBar stateStore={stateStore} />
            <div className="photo-portfolio">
                <Routes>
                    <Route
                        path="/"
                        element={<Categories stateStore={stateStore} />}
                    />
                    <Route
                        path="page/:pageId"
                        element={<Page key={pageId} stateStore={stateStore} />}
                    />
                    <Route
                        path="category/:categoryId"
                        element={<Category key={categoryId} stateStore={stateStore} />}
                    >
                        <Route
                            path="post/:postId"
                            element={<Post stateStore={stateStore} />}
                        />
                    </Route>
                </Routes>
            </div>
            <Menu stateStore={stateStore} />
            <Footer />
        </div>
    );
};

PhotoPortfolio.displayName = 'PhotoPortfolio';

export default observer(PhotoPortfolio);
