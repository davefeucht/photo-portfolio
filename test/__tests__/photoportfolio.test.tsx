import { render, screen, waitFor } from '@testing-library/react';
import PhotoPortfolio from 'components/PhotoPortfolio/PhotoPortfolio';
import * as React from 'react';
import { MemoryRouter } from 'react-router-dom';
import StateStore from 'StateStore/store';

import { categories } from '../data/testData';

jest.mock('../../utils/WordpressAPI');

test('PhotoPortfolio displays', async () => {
    const store = new StateStore();
    store.setCategoryList(categories)
    const { container } = render(
        <MemoryRouter initialEntries={['/']}>
            <PhotoPortfolio stateStore={store} />
        </MemoryRouter>
    );

    await waitFor(() => expect(screen.getAllByText(categories[0].name).length).not.toEqual(0));
    expect(container.firstChild).toMatchSnapshot();
});
