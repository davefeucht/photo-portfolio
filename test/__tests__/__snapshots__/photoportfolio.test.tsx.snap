// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoPortfolio displays 1`] = `
<div
  class="app"
>
  <div
    class="title-bar"
  >
    <div
      class="hamburger-button closed"
    >
      <img
        src="./assets/images/hamburger_icon.svg"
        style="border: 0px;"
      />
    </div>
    <div>
      <h1>
        Through a Pinhole
      </h1>
    </div>
  </div>
  <div
    class="photo-portfolio"
  >
    <div
      class="category"
    >
      <div
        class="section-header"
      >
        <div
          class="title"
        />
      </div>
      <div
        class="posts"
      >
        <div
          class="pagination-navigation"
        >
          <div
            class="pagination-navigation__content"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="menu closed"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          <div
            class="menu-link"
          >
            Home
          </div>
        </a>
      </li>
      <li>
        Categories
        <ul />
      </li>
      <li>
        Info
        <ul />
      </li>
    </ul>
  </div>
  <div
    class="footer"
  >
    Theme Copyright @2023 Dave Feucht
  </div>
</div>
`;
